charmcraft/extensions/gunicorn.py

```diff
@@ -0,0 +1,241 @@
+    @staticmethod
+    @override
+    def is_experimental(base: tuple[str, ...] | None) -> bool:  # noqa: ARG004
+        """Check if the extension is in an experimental state."""
+        return True
+
+    @property
+    @abc.abstractmethod
+    def framework(self) -> str:
+        """Return the name of the WSGI framework."""
+
+    def _get_nested(self, obj: dict, path: str) -> dict:
+        """Get a nested object using a path (a dot-separated list of keys)."""
+        for key in path.split("."):
+            obj = obj.get(key, {})
+        return obj
+
+    def check_input(self) -> None:
+        """Check if the extension is applicable for user input charmcraft project file."""
+        for incompatible_field in ("devices", "extra-bindings", "storage"):
+            if incompatible_field in self.yaml_data:
+                raise ExtensionError(
+                    f"the {self.framework}-framework extension is not compatible with the "
+                    f"field {incompatible_field!r}"
+                )
+        root_snippet = self._get_root_snippet()
+        for protected in ("type", "assumes", "containers", "resources", "peers"):
+            if (
+                protected in self.yaml_data
+                and self.yaml_data[protected] != root_snippet[protected]
+            ):
+                raise ExtensionError(
+                    f"{protected!r} in charmcraft.yaml conflicts with a reserved field "
+                    f"in the {self.framework}-framework extension, please remove it."
+                )
+        for merging in ("actions", "requires", "provides", "config.options"):
+            user_provided: dict[str, Any] = self._get_nested(self.yaml_data, merging)
+            if not user_provided:
+                continue
+            overlap = user_provided.keys() & self._get_nested(root_snippet, merging).keys()
+            if overlap:
+                raise ExtensionError(
+                    f"overlapping keys {overlap} in {merging} of charmcraft.yaml "
+                    f"which conflict with the {self.framework}-framework extension, "
+                    "please rename or remove it"
+                )
+        for config in self._get_nested(self.yaml_data, "config.options"):
+            for reserved_config_prefix in ("webserver-", f"{self.framework}-"):
+                if config[: len(reserved_config_prefix)] == reserved_config_prefix:
+                    raise ExtensionError(
+                        f"config.options {config!r} starts with {self.framework}-framework"
+                        f" reserved configuration prefix {reserved_config_prefix!r}, "
+                        "please rename or remove it"
+                    )
+
+    def _get_root_snippet(self) -> dict[str, Any]:
+        """Return the root snippet to be merged into the user charmcraft.yaml.
+
+        This method differs from get_root_snippet because it doesn't perform any check.
+        """
+        return {
+            "type": "charm",
```

Comment:
I would rather keep this explicit in all `charmcraft.yaml` files unless we have direction from above otherwise.

Reply:
Gotcha, remove the `type` from the root snippet and added a new check to make sure it's `charm`. Thanks